[docs] best practices for upgrading zenml server (#3087)
zenml-io/zenml: docs/book/how-to/manage-the-zenml-server/using-zenml-server-in-prod.md

---
description: >
  Learn about best practices for using ZenML server in production environments.
---

# Using ZenML server in production

> Setting up a ZenML server is pretty straightforward. However, most people have to move ahead of day zero operations and in such cases, it helps to learn best practices around setting up your ZenML server in a production-ready way. This guide encapsulates all the tips and tricks we've learned ourselves and from working with people who use ZenML in production environments. Following are some of the best practices we recommend.

Reviewer: ```suggestion
Setting up a ZenML server is pretty straightforward. However, most people have to move ahead of so-called 'day zero' operations and in such cases, it helps to learn best practices around setting up your ZenML server in a production-ready way. This guide encapsulates all the tips and tricks we've learned ourselves and from working with people who use ZenML in production environments. Following are some of the best practices we recommend.
```

> Setting up a ZenML server is pretty straightforward. However, most people have to move ahead of day zero operations and in such cases, it helps to learn best practices around setting up your ZenML server in a production-ready way. This guide encapsulates all the tips and tricks we've learned ourselves and from working with people who use ZenML in production environments. Following are some of the best practices we recommend.

Reviewer 2: Dont say "its striaghtforward" and then have a huge guide :-) no need to take emotional opinions I think

Author: ah, i agree. i wanted it to mean that just deploying vanilla zenml server is easy but then managing and maintaining it is hard. let me rephrase that

{% hint style="info" %}
If you are using ZenML Pro, you don't have to worry about any of these. We have got you covered!
You can sign up for a free trial [here](https://cloud.zenml.io).
{% endhint %}

## Autoscaling replicas

In production, you often have to run bigger and longer running pipelines that might strain your server's resources. It is a good idea to set up autoscaling for your ZenML server so that you don't have to worry about your pipeline runs getting interrupted or your Dashboard slowing down due to high traffic.

> How you do it depends greatly on the environment you have deployed your ZenML server in. Below are some commmon deployment options and how to set up autoscaling for them.

Reviewer: ```suggestion
How you do it depends greatly on the environment in which you have deployed your ZenML server. Below are some common deployment options and how to set up autoscaling for them.
```

{% endtab %}
{% endtabs %}

## High connection pool values

One other way to improve the performance of your ZenML server is to increase the number of threads that your server process uses, provided that you have hardware that can support it.

You can control this by setting the `zenml.threadPoolSize` value in the ZenML Helm chart values. For example:

```yaml
zenml:
  threadPoolSize: 100
```

By default, it is set to 40. If you are using any other deployment option, you can set the `ZENML_SERVER_THREAD_POOL_SIZE` environment variable to the desired value.

Once this is set, you should also modify the `zenml.database.poolSize` and `zenml.database.maxOverflow` values to ensure that the ZenML server workers do not block on database connections (i.e. the sum of the pool size and max overflow should be greater than or equal to the thread pool size). If you manage your own database, ensure these values are set appropriately.

## Scaling the backing database

An important component of the ZenML server deployment is the backing database. When you start scaling your ZenML server instances, you will also need to scale your database to avoid any bottlenecks.

> We would recommend starting out with a simple database instance and then monitoring it to decide if it needs scaling. Some common metrics to look out for:

Reviewer: ```suggestion
We would recommend starting out with a simple (single) database instance and then monitoring it to decide if it needs scaling. Some common metrics to look out for:
```

{% endtab %}
{% tab title="ECS" %}
On ECS, you can utilize the [CloudWatch integration](https://docs.aws.amazon.com/AmazonECS/latest/developerguide/cloudwatch-metrics.html) to monitor your ZenML server.

In the "Health and metrics" section of your ECS console, you should see metrics pertaining to your ZenML service like CPU utilization and Memory utilization.

![Image showing CPU utilization ECS](../../.gitbook/assets/ecs_cpu_utilization.png)

{% endtab %}
{% tab title="Cloud Run" %}
In Cloud Run, you can utilize the [Cloud Monitoring integration](https://cloud.google.com/run/docs/monitoring) to monitor your ZenML server.

The "Metrics" tab in the Cloud Run console will show you metrics like Container CPU utilization, Container memory utilization, and more.

![Image showing metrics in Cloud Run](../../.gitbook/assets/cloudrun_metrics.png)

{% endtab %}
{% endtabs %}

## Backups

The data in your ZenML server is critical as it contains your pipeline runs, stack configurations, and other important information. It is, therefore, recommended to have a backup strategy in place to avoid losing any data.

> Some common strategies include:
> - Setting up automated backups with a good retention period (say 30 days).

Reviewer 2: ```suggestion
Some common strategies include:

- Setting up automated backups with a good retention period (say 30 days).
```

In the "Health and metrics" section of your ECS console, you should see metrics pertaining to your ZenML service like CPU utilization and Memory utilization.

![Image showing CPU utilization ECS](../../.gitbook/assets/ecs_cpu_utilization.png)

{% endtab %}
{% tab title="Cloud Run" %}
In Cloud Run, you can utilize the [Cloud Monitoring integration](https://cloud.google.com/run/docs/monitoring) to monitor your ZenML server.

The "Metrics" tab in the Cloud Run console will show you metrics like Container CPU utilization, Container memory utilization, and more.

![Image showing metrics in Cloud Run](../../.gitbook/assets/cloudrun_metrics.png)

{% endtab %}
{% endtabs %}

## Backups

The data in your ZenML server is critical as it contains your pipeline runs, stack configurations, and other important information. It is, therefore, recommended to have a backup strategy in place to avoid losing any data.

Some common strategies include:

- Setting up automated backups with a good retention period (say 30 days).
- Periodically exporting the data to an external storage (e.g. S3, GCS, etc.).
- Manual backups before upgrading your server to avoid any problems.

> 
> 

Reviewer 2: ```suggestion
```
